src/git.rs

```diff
@@ -27,6 +27,18 @@ pub fn repo_hash(path: impl AsRef<Path>) -> Option<String> {
     }
 }
 
+pub(crate) fn repo_remote_url(path: impl AsRef<Path>) -> Option<String> {
+    let get_url = |remote: &str| {
+        let output = git_command(path.as_ref(), &["remote", "get-url", remote]).ok()?;
+        // XXX: Use `.then_some` when it is stabilized
+        output.status.success().then(|| output)
+    };
+    let output = get_url("upstream").or_else(|| get_url("origin"))?;
```

Comment:
That really doesn't sound like a good idea... The best way would be to check for the default remote and use this one instead. I don't like much harcoded things that depends on users' config.

Reply:
Is there such a thing as a "default remote"? I know every branch can have its own "tracking remote" but that's different. In any case it is more likely to have ones fork as default remote, as that's what you'll end up pushing to. Maybe looking at the pull-remote for `master` is a good idea?

Anyway, let me quote a piece from the PR description again:

> Consensus shows that these repos are either directly cloned through submodules with a default remote named `origin` or a manual rename to `upstream` with the personal remote (from which PRs are commonly created) named `origin`.

It seems like that's the best we can do, bar hardcoding the gstreamer URL (and future ones) in `gir` or not providing a URL at all.

Reply:
Why is `upstream` the default then? For git submodules, it seems like you instead want to use `origin`. And in case none of the two remotes are there, we are back to the original situation. To prevent that, we could use `git remote -v` (which could actually prevent to run multiple `git` commands since all remotes are listed at once).

And finally: if this is a `git@` URL, it's definitely not very useful, which is why I didn't add it in the first place. :-/

Reply:
You can get something like that via `git branch -vv`. Not sure how to properly read what `git branch --set-upstream-to` points at.

However I think this all does not work with detached branches, like what `git submodule` would create. I don't think git actually knows what the "canonical" origin is where the data came from.

I guess a more correct way for the hacks you and @MarijnS95 are doing would be to set a different push and fetch URL for the `origin`, and leave the `fetch` one intact, and use `origin` the way how it's supposed to be used ;)

See e.g.

```
$ git remote -v
origin	https://gitlab.freedesktop.org/gstreamer/gstreamer.git (fetch)
origin	git@gitlab.freedesktop.org:slomo/gstreamer.git (push)
```

Reply:
> Why is `upstream` the default then?

Since both [you](https://github.com/gtk-rs/gir/pull/1041#issuecomment-793849521) and me seem to have renamed the "upstream" repo to that, and it seems to be a common pattern when working with a fork to create PRs from (or `origin`+`fork`, which this also works for). Indeed, ssh links make little sense and I have it cloned as `https` because it is a public repo; I guess you are pushing (new branches) directly to it every once in a while? A https-fetch and ssh-push URL as Sebastian make most sense here, but both to your own fork.

> (which could actually prevent to run multiple `git` commands since all remotes are listed at once).

I prefer running multiple commands providing just the information we need rather than attempting to parse human-readable output that is not deemed stable to parse in the first place.

> You can get something like that via `git branch -vv`. Not sure how to properly read what `git branch --set-upstream-to` points at.

There are multiple tricks: https://stackoverflow.com/a/9753364/2844473

> However I think this all does not work with detached branches, like what `git submodule` would create. I don't think git actually knows what the "canonical" origin is where the data came from.

We'd have to find a `master` branch; does it create one and track the initial `origin` upon clone/checkout?

> I guess a more correct way for the hacks you and @MarijnS95 are doing would be to set a different push and fetch URL for the `origin`, and leave the `fetch` one intact, and use `origin` the way how it's supposed to be used ;)

That afaik has complications when pushing to and pulling from my own remote :grimacing: 

Fwiw `get-url` returns the fetch-url by default, unless `--push` is passed.

Reply:
For submodules, just let git do its thing (i.e. it will become `origin`). Who knows what other side effects changing the `origin` remote has with the other `git submodule` commands. It's fragile enough even without such things :)

Personally I have a separate clone of the submodules in a different place, which I use for changes to them. For other repos I have a setup like above, and my fork is write-only.

Reply:
I really think that overall we're trying to do clever things that are way too focus on the existing code. And even that isn't working out so well. Not sure this is worth it...

Reply:
I would just use `origin` here and if someone manually broke the submodule configuration then that's their problem ;)

Reply:
> For submodules, just let git do its thing (i.e. it will become `origin`). Who knows what other side effects changing the `origin` remote has with the other `git submodule` commands. It's fragile enough even without such things :)

No need to mess with submodules from `gir` itself (unless you want to query whether the specified directory is a submodule and extract a url from that), and anyway preferring `upstream` over `origin` still works just as well on the `origin`-only checkout doesn't it?

> Personally I have a separate clone of the submodules in a different place, which I use for changes to them. For other repos I have a setup like above, and my fork is write-only.

I find that massively inconvenient if for example working on the combination of gir, gstreamer and gir-files in one spot - all changes have to be in the same place in order to even work.

> I would just use `origin` here and if someone manually broke the submodule configuration then that's their problem ;)

Again, `origin` is still supported just as well; does it really hurt to prefer `upstream`'s url over that?

Reply:
> does it really hurt to prefer upstream's url over that?

I'm fine with your change

Reply:
My compromise proposal here would be to always get the submodule URL via `.gitmodules`. That should always be correct

Reply:
Or maybe if @GuillaumeGomez insists on having the URL configurable in `Gir.toml`, we shouldn't use any git submodules at all and have `gir` clone the repo / commit that is configured in `Gir.toml` for each of the directories?

Reply:
As discussed with @GuillaumeGomez yesterday, we would go with looking at the git configuration directly for the submodule. So basically for the submodule at path `gir-files` we would get the submodule name

```sh
# note the escaping for the backslashes so they end up in the command and the dot is not considered as an any-character
git config --file .gitmodules --get-regexp 'submodule\\..*\\.path' '^gir-files$'
```

This prints `submodule.gir-files.path gir-files`, the first `gir-files` is the submodule name.

Then

```sh
git config --file .gitmodules --get-regexp 'submodule\\.gir-files\\.url'
```

This prints `submodule.gir-files.url https://github.com/gtk-rs/gir-files` and there's the URL.

For this URL to be wrong for the submodule, you'd have to actually change the submodule configuration and then it would also show up as a change in the main git repository.

It would be resilient to whatever hacks you're doing inside the submodule repository itself.

Reply:
@MarijnS95 What do you think?

Reply:
@sdroege I think that I may not have enough time to fully dig into the suggested change, but will try...

I guess that's a smart way, yes. First map the path to a submodule, then back to a URL. Can we do it in a more rusty way though, ie. https://docs.rs/git2/0.13.17/git2/struct.Config.html? Perhaps we can convert the whole `git` command mangling over to a proper git wrapper while at it.

Reply:
That's where we were at some point, see https://github.com/gtk-rs/gir/pull/855 . I don't mind but some people are allergic to dependencies :)

For this case it seems the parsing of the output is easy enough with the `&str` iterators and functions like `strip_prefix()`, `split_once()`, etc.

Reply:
Yikes, maybe that is resolved by now by trimming down the `-sys` crate in some way? Surprisingly `gstreamer-rs` with all its autogenerated code compiles even faster than that :joy:

Reply:
It seems easy enough for now to paste the git output in this case. I'd worry about that in the future if something complicated happens :)

Reply:
@sdroege It is finally weekend and I finally have some time :partying_face:! `split_once` would have been awesome but that's still unstable :rage:, but there's an even better solution: `--name-only`! (For some reason there is no `--value-only` equivalent when using `--get-regexp`, fortunately we don't need it in the end).

The end result is pretty elegant, if I may say so myself (bar some string manipulation to trim `\0` etc, improvement suggestions are welcome). I'll leave some self-review comments.

Reply:
Oh I thought `split_once` was finally stable in 1.51. It's 1.52 apparently, oh well!

Reply:
Now that `split_once` lands today, anywhere we can use this? :P